File: C/Baekjoon/11000/11438.cpp
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    int T, N, M, a, b, c, diff, j, root;
    cin >> T;

    while (T--)
    {
        cin >> N;

        vector<int> arr[N + 1];
        int parent[N + 1][15] = {}, depth[N + 1] = {};

        for (int i = 0; i < N - 1; i++)
        {
            cin >> a >> b;
            parent[b][0] = a;
            arr[a].push_back(b);
        }

        for (int i = 1; i <= N; i++)
        {
            if (!parent[i][0])
                root = i;
        }

        stack<int> st;
        st.push(root);

        while (!st.empty())
        {
            a = st.top();
            st.pop();

            for (int i : arr[a])
            {
                depth[i] = depth[a] + 1;
                for (int j = 1; 1 << j <= N; j++)
                {
                    if (!parent[i][j - 1])
                        break;
                    parent[i][j] = parent[parent[i][j - 1]][j - 1];
                }

                st.push(i);
            }
        }

        cin >> a >> b;
        if (depth[a] < depth[b])
        {
            c = b;
            b = a;
            a = c;
        }

        diff = depth[a] - depth[b];
        j = 0;

        while (diff)
        {
            if (diff % 2)
                a = parent[a][j];
            j++;
            diff /= 2;
        }

        for (int j = 14; j >= 0; j--)
        {
            if (parent[a][j] != parent[b][j])
            {
                a = parent[a][j];
                b = parent[b][j];
            }
        }
        if (a != b)
            a = parent[a][0];

        cout << a << '\n';
    }
}

File: PS/Baekjoon/11000/11438.py
from sys import stdin
from collections import deque

N = int(stdin.readline())
arr = [[] for i in range(N + 1)]

for i in range(N - 1):
    A, B = map(int, stdin.readline().rstrip().split())
    arr[A].append(B)
    arr[B].append(A)

st = deque([1])
checked = [False for i in range(N + 1)]
checked[1] = True

depth = [0 for i in range(N + 1)]
parent = [[0 for j in range(18)] for i in range(N + 1)]

while len(st):
    p = st.pop()
    for q in arr[p]:
        if checked[q]:
            continue

        checked[q] = True
        parent[q][0] = p
        depth[q] = depth[p] + 1

        k = 1
        while 2**k <= N:
            if not parent[q][k - 1]:
                break
            parent[q][k] = parent[parent[q][k - 1]][k - 1]
            k += 1

        st.append(q)

top = [1]
for i in range(17):
    top.append(top[-1] * 2)

M = int(stdin.readline())
for i in range(M):
    A, B = map(int, stdin.readline().rstrip().split())
    if depth[B] > depth[A]:
        A, B = B, A

    diff = depth[A] - depth[B]
    for j in range(17, -1, -1):
        if top[j] <= diff:
            diff -= top[j]
            A = parent[A][j]

    for j in range(17, -1, -1):
        if parent[A][j] != parent[B][j]:
            A, B = parent[A][j], parent[B][j]
    if A != B:
        A = parent[A][0]

    print(A)
